File: naked-generator/Primary/c/acunsteer.c
#include <string.h>
#include <assert.h>
#include <stdio.h>

#include "acunsteer.h"

void serialize_HV_VOLTAGE(HV_VOLTAGE* hv_voltage, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(HV_VOLTAGE));
	memcpy(buffer, hv_voltage, sizeof(STEER_STATUS));
}
void serialize_HV_CURRENT(HV_CURRENT* hv_current, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(HV_CURRENT));
	memcpy(buffer, hv_current, sizeof(STEER_STATUS));
}
void serialize_HV_TEMP(HV_TEMP* hv_temp, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(HV_TEMP));
	memcpy(buffer, hv_temp, sizeof(STEER_STATUS));
}
void serialize_HV_ERROR(HV_ERROR* hv_error, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(HV_ERROR));
	memcpy(buffer, hv_error, sizeof(STEER_STATUS));
}

void deserialize_HV_VOLTAGE(uint8_t* buffer, size_t buf_len, HV_VOLTAGE* hv_voltage) {
	assert(buf_len >= sizeof(HV_VOLTAGE));
	memcpy(hv_voltage, buffer, sizeof(STEER_STATUS));
}
void deserialize_HV_CURRENT(uint8_t* buffer, size_t buf_len, HV_CURRENT* hv_current) {
	assert(buf_len >= sizeof(HV_CURRENT));
	memcpy(hv_current, buffer, sizeof(STEER_STATUS));
}
void deserialize_HV_TEMP(uint8_t* buffer, size_t buf_len, HV_TEMP* hv_temp) {
	assert(buf_len >= sizeof(HV_TEMP));
	memcpy(hv_temp, buffer, sizeof(STEER_STATUS));
}
void deserialize_HV_ERROR(uint8_t* buffer, size_t buf_len, HV_ERROR* hv_error) {
	assert(buf_len >= sizeof(HV_ERROR));
	memcpy(hv_error, buffer, sizeof(STEER_STATUS));
}

File: naked-generator/Primary/py/acunsteer.py
from enum import Enum
from struct import pack, unpack
from collections import namedtuple


HV_VOLTAGE = namedtuple('HV_VOLTAGE', 'pack_voltage bus_voltage max_cell_voltage min_cell_voltage')
HV_VOLTAGE_schema = '<HHHH'

HV_CURRENT = namedtuple('HV_CURRENT', 'current power')
HV_CURRENT_schema = '<Hc'

HV_TEMP = namedtuple('HV_TEMP', 'average_temp max_temp min_temp')
HV_TEMP_schema = '<HHH'

HV_ERROR = namedtuple('HV_ERROR', 'error_code error_index active')
HV_ERROR_schema = '<BBB'


def serialize_HV_VOLTAGE(struct: HV_VOLTAGE) -> bytearray:
	return pack(HV_VOLTAGE, *tuple(struct))


def serialize_HV_CURRENT(struct: HV_CURRENT) -> bytearray:
	return pack(HV_CURRENT, *tuple(struct))


def serialize_HV_TEMP(struct: HV_TEMP) -> bytearray:
	return pack(HV_TEMP, *tuple(struct))


def serialize_HV_ERROR(struct: HV_ERROR) -> bytearray:
	return pack(HV_ERROR, *tuple(struct))


def deserialize_HV_VOLTAGE(buffer: bytearray) -> HV_VOLTAGE:
	return HV_VOLTAGE._make(unpack(HV_VOLTAGE_schema, buffer))


def deserialize_HV_CURRENT(buffer: bytearray) -> HV_CURRENT:
	return HV_CURRENT._make(unpack(HV_CURRENT_schema, buffer))


def deserialize_HV_TEMP(buffer: bytearray) -> HV_TEMP:
	return HV_TEMP._make(unpack(HV_TEMP_schema, buffer))


def deserialize_HV_ERROR(buffer: bytearray) -> HV_ERROR:
	return HV_ERROR._make(unpack(HV_ERROR_schema, buffer))


